### PR TITLE
sd: refactor the SD's backend interface

### PR DESCRIPTION
**Pull request by arogge**
This changes the SD Backend and Device interface in a way that is supposed to make it easier to add new backends.

* file device is now also built and loaded dynamically
* `Device Type` setting will accept any string
* presence of a configured storage backend checked after loading the configuration
* systemtests fail early if a daemon config doesn't pass a test with `-t`
* Remove `IsFile()` and `IsFifo()` from `Device` (in favor of new checks)
* Increase CMake required version to 3.17
* static linking of chunked, droplet and gentape into the backends that use them

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

##### General
- [x] PR name is meaningful
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Check backport line
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Separate commit for CHANGELOG.md ("update CHANGELOG.md"). The PR number is correct.
- [x] Merge corresponding PR in git.bareos.com/ci-scripts

##### Source code quality

- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR
- [x] `bareos-check-sources --since-merge` does not report any problems

### CODE REVIEW

File: core/src/lib/plugin_registry.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/*
```
Comment (pstorz) [inline core/src/lib/plugin_registry.h:1]: Maybe better call it "backend registry" to avoid confusion with fd, sd and dir plugins?

**Review by pstorz — CHANGES_REQUESTED** [review]
_Do we still want to ship libbareossd-chunked.so and libbareossd-gentape.so?_
**No, we want to link chunked and gentape statically into the backends (i.e. droplet and tape) instead of shipping them as shared libraries.**

File: core/src/stored/device_resource.h [hunk 2]
```diff
@@ -43,7 +43,7 @@ class DeviceResource : public BareosResource {
   char* changer_command;        /**< Changer command  -- external program */
   char* alert_command;          /**< Alert command -- external program */
   char* spool_directory;        /**< Spool file directory */
-  DeviceType dev_type;          /**< device type */
+  std::string dev_type;         /**< device type */
```
Comment (arogge) [inline core/src/stored/device_resource.h:46]:
```suggestion
  std::string device_type;
```

File: core/src/lib/plugin_registry.h (hunk 1 above)
Comment (arogge) [reply core/src/lib/plugin_registry.h]: renamed it to ImplementationFactory

File: core/CMakeLists.txt [hunk 3]
```diff
@@ -30,6 +30,7 @@ endif()
 
 cmake_policy(SET CMP0054 NEW)
 cmake_policy(SET CMP0057 NEW)
+cmake_policy(SET CMP0079 NEW)
```
Comment (pstorz) [inline core/CMakeLists.txt:33]:
```suggestion
cmake_policy(SET CMP0054 NEW) # Only interpret if() arguments as variables or keywords when unquoted.
cmake_policy(SET CMP0057 NEW) # Support new if() IN_LIST operator.
cmake_policy(SET CMP0079 NEW) # target_link_libraries() allows use with targets in other directories.
```

Comment (pstorz) [reply core/CMakeLists.txt:33]: Also I think we should request  `cmake_minimum_required(VERSION 3.17)`

**Review by pstorz — CHANGES_REQUESTED** [review]
Just minimal changes, please check my comments.

File: core/CMakeLists.txt (hunk 3 above)
Comment (arogge) [reply core/CMakeLists.txt:33]: we do that in the toplevel CMakeLists.txt

Comment (arogge) [reply core/CMakeLists.txt]: I unified cmake_policy to require 3.17 and enable all policies up to and including 3.19.
We also require CMake 3.17 now and don't use `cmake_minimum_required()` everywhere anymore.

**Review by pstorz — APPROVED** [review]
Good work!

**Pull request merged**